File: deposit_wallets/5_init_clob.py
"""
5_init_clob.py — Init ClobClient V2 with deposit wallet (signatureType=3 / POLY_1271).

Run:
    python 5_init_clob.py

Requires: PRIVATE_KEY, DEPOSIT_WALLET, CLOB_V2_BASE_URL in .env
signatureType=3 (POLY_1271): the deposit wallet is both the funder and the
order maker/signer. The EOA (or session signer) signs ERC-7739-wrapped payloads
validated by ERC-1271 on the deposit wallet contract.
"""

import os
import sys
import requests
sys.path.insert(0, os.path.dirname(__file__))

from dotenv import load_dotenv
from py_clob_client_v2 import ClobClient, SignatureTypeV2

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

PRIVATE_KEY    = os.environ["PRIVATE_KEY"]
DEPOSIT_WALLET = os.environ["DEPOSIT_WALLET"]
CLOB_HOST      = os.environ.get("CLOB_V2_BASE_URL", "https://clob.polymarket.com")
EVENT_ID       = "73106"  # replace with your market event ID

print(f"CLOB           : {CLOB_HOST}")
print(f"Deposit wallet : {DEPOSIT_WALLET}\n")

# ── Step 1: Derive CLOB API credentials (deterministic, tied to deposit wallet) ─
temp_client = ClobClient(
    CLOB_HOST,
    chain_id=137,
    key=PRIVATE_KEY,
    signature_type=SignatureTypeV2.POLY_1271,
    funder=DEPOSIT_WALLET,
)
creds = temp_client.create_or_derive_api_key()
print(f"API key : {creds.api_key} ✓")

# ── Step 2: Authenticated client ────────────────────────────────────────────────
client = ClobClient(
    CLOB_HOST,
    chain_id=137,
    key=PRIVATE_KEY,
    creds=creds,
    signature_type=SignatureTypeV2.POLY_1271,
    funder=DEPOSIT_WALLET,
)
print("ClobClient V2 (deposit wallet / POLY_1271) initialized ✓\n")

# ── Step 3: Fetch market info ────────────────────────────────────────────────────
event   = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id = event["markets"][0]["conditionId"]
info    = client.get_clob_market_info(cond_id)
tokens  = info.get("t", [])
yes_id  = next((t["t"] for t in tokens if t["o"] == "Yes"), "")

print(f"Condition ID : {cond_id}")
print(f"Tick size    : {info.get('mts', '?')}")
print(f"Neg risk     : {info.get('nr', False)}")
print(f"YES token    : {yes_id}")

# ── Step 4: Order book ───────────────────────────────────────────────────────────
book = client.get_order_book(yes_id)
bids = sorted(book["bids"], key=lambda b: float(b["price"]), reverse=True)
asks = sorted(book["asks"], key=lambda a: float(a["price"]))
mid  = (float(bids[0]["price"]) + float(asks[0]["price"])) / 2 if bids and asks else None

print(f"\nOrder book — {len(bids)} bids / {len(asks)} asks ✓")
for a in reversed(asks[:3]):
    print(f"  ask  ${float(a['price']):.4f}  sz:{a['size']}")
if mid:
    print(f"  ── mid ──  ${mid:.4f}")
for b in bids[:3]:
    print(f"  bid  ${float(b['price']):.4f}  sz:{b['size']}")

File: deposit_wallets_ts/5_init_clob.ts
/**
 * 5_init_clob.ts — Init ClobClient with POLY_1271 (deposit wallet), derive API key,
 *                  fetch market info + order book.
 *
 * Run: npx tsx 5_init_clob.ts
 * Requires: PRIVATE_KEY, DEPOSIT_WALLET, CLOB_V2_BASE_URL in .env
 */

import "dotenv/config";
import { makeClient } from "./_clob.js";

// JD Vance 2028
const COND_ID      = "0x7ad403c3508f8e3912940fd1a913f227591145ca0614074208e0b962d5fcc422";
const YES_TOKEN_ID = "16040015440196279900485035793550429453516625694844857319147506590755961451627";

const client = await makeClient();

const info = await client.getClobMarketInfo(COND_ID);
console.log(`Market info : tick=${info.mts}  negRisk=${info.nr}  tokens=${JSON.stringify(info.t?.map((t: any) => t.o))}`);

const book = await client.getOrderBook(YES_TOKEN_ID);
const bids = ((book as any).bids ?? []).sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
const asks = ((book as any).asks ?? []).sort((a: any, b: any) => parseFloat(a.price) - parseFloat(b.price));

console.log(`\nOrder book (YES):`);
console.log(`  Best bid : ${bids[0]?.price ?? "n/a"}  size=${bids[0]?.size ?? "n/a"}`);
console.log(`  Best ask : ${asks[0]?.price ?? "n/a"}  size=${asks[0]?.size ?? "n/a"}`);
console.log(`\n✓ CLOB client initialised successfully`);
